lib/elixir/lib/calendar/iso.ex

```diff
@@ -1455,6 +1455,134 @@ defmodule Calendar.ISO do
     {days, {@parts_per_day - 1, @parts_per_day}}
   end
 
+  @doc """
+  Shifts date by Calendar.Duration units according to its calendar.
+
+  Available units are: `:year, :month, :week, :day, :hour, :minute, :second, :microsecond`.
+
+  When shifting by month:
+  - it will shift to the current day of a month
+  - when the current day does not exist in a month, it will shift to the last day of a month
+
+  ## Examples
+
+      iex> Calendar.ISO.shift_date(2016, 1, 3, month: 2)
+      {2016, 3, 3}
+      iex> Calendar.ISO.shift_date(2016, 2, 29, month: 1)
+      {2016, 3, 29}
+      iex> Calendar.ISO.shift_date(2016, 1, 31, month: 1)
+      {2016, 2, 29}
+      iex> Calendar.ISO.shift_date(2016, 1, 31, year: 4, day: 1)
+      {2020, 2, 1}
+  """
+  @spec shift_date(year(), month(), day(), [Calendar.Duration.duration_unit()]) ::
+          {year, month, day}
+  @impl true
+  def shift_date(year, month, day, duration_units) do
+    {year, month, day, _, _, _, _} =
+      shift_naive_datetime(year, month, day, 0, 0, 0, {0, 0}, duration_units)
+
+    {year, month, day}
+  end
+
+  @doc """
+  Shifts naive datetime by Calendar.Duration units according to its calendar.
+
+  Available units are: `:year, :month, :week, :day, :hour, :minute, :second, :microsecond`.
+
+  When shifting by month:
+  - it will shift to the current day of a month
+  - when the current day does not exist in a month, it will shift to the last day of a month
+
+  ## Examples
+
+      iex> Calendar.ISO.shift_naive_datetime(2016, 1, 3, 0, 0, 0, {0, 0}, hour: 1)
+      {2016, 1, 3, 1, 0, 0, {0, 0}}
+      iex> Calendar.ISO.shift_naive_datetime(2016, 1, 3, 0, 0, 0, {0, 0}, hour: 30)
+      {2016, 1, 4, 6, 0, 0, {0, 0}}
+      iex> Calendar.ISO.shift_naive_datetime(2016, 1, 3, 0, 0, 0, {0, 0}, microsecond: 100)
+      {2016, 1, 3, 0, 0, 0, {100, 6}}
+  """
+  @spec shift_naive_datetime(year, month, day, hour, minute, second, microsecond, [
+          Calendar.Duration.duration_unit()
+        ]) :: {year, month, day, hour, minute, second, microsecond}
+  @impl true
+  def shift_naive_datetime(year, month, day, hour, minute, second, microsecond, duration_units) do
+    Enum.reduce(duration_units, {year, month, day, hour, minute, second, microsecond}, fn
+      {_opt, 0}, naive_datetime -> naive_datetime
+      {:year, value}, naive_datetime -> shift_years(naive_datetime, value)
```

Comment:
The general consensus is to indeed add years, then months, then days, etc, but at least Java adds years and months together: https://docs.oracle.com/javase/8/docs/api/java/time/Period.html#addTo-java.time.temporal.Temporal-

Of course, that means the result is different but it seems logical to me. So I think the correct implementation here would be to first collapse all dimensions in two: month and seconds (or microseconds) and then perform two operations only: one adding the months and the other adding the seconds (or microseconds).

Reply:
Yeah, you will end-up with two shift calls:

  * `shift_unit(ndt, value, :second | :microsecond)`
  * `shift_month(ndt, months)`

In practice, `shift_unit` can accept all of the same units as `NaiveDateTime.add` and negotiate the precision as in that function.

Reply:
Right! that's what i am doing now, in a possibly too verbose manner :)